File: ray_tracing/material.py
from ray_tracing.ray import Ray
from ray_tracing.vector import random_in_hemisphere, random_unit, reflect, vector


class Material(object):
    def reflect(self, ray, t, normal):
        raise NotImplementedError

    def emit(self):
        raise NotImplementedError


class Metal(Material):
    def __init__(self, color):
        self.color = color

    def reflect(self, ray, t, normal):
        reflected = reflect(ray.direction, normal)
        reflected = Ray(ray.position_at(t), reflected)

        if reflected.direction.dot(normal) <= 0:
            return None

        return Reflection(reflected, self.color)

    def emit(self):
        return vector(0, 0, 0)


class Diffuse(Material):
    def __init__(self, color, mode="hemi"):
        self.color = color
        self.mode = mode

    def reflect(self, ray, t, normal):
        if self.mode == "sphere":
            return self.reflect_sphere(ray, t, normal)
        elif self.mode == "hemi":
            return self.reflect_hemi(ray, t, normal)
        else:
            raise ValueError("invalid mode {}".format(self.mode))

    def reflect_sphere(self, ray, t, normal):
        ray = Ray(ray.position_at(t), normal + random_unit())

        return Reflection(ray, self.color)

    def reflect_hemi(self, ray, t, normal):
        ray = Ray(ray.position_at(t), random_in_hemisphere(normal))

        return Reflection(ray, self.color)

    def emit(self):
        return vector(0, 0, 0)


class Light(Material):
    def __init__(self, color):
        self.color = color

    def reflect(self, ray, t, normal):
        return None

    def emit(self):
        return self.color


class Reflection(object):
    def __init__(self, ray, attenuation):
        self.ray = ray
        self.attenuation = attenuation

File: ray_tracing_rust/src/material.rs
use crate::color::Color;
use crate::ray::Ray;
use crate::reflection::Reflection;
use crate::vector::Vector3;

pub trait Material {
    fn reflect(&self, ray: &Ray, t: f32, normal: Vector3) -> Option<Reflection>;
    fn emit(&self) -> Color;
}

pub struct Light {
    color: Color,

    _secret: (),
}

impl Light {
    pub fn new(color: Color) -> Self {
        Self { color, _secret: () }
    }
}

impl Material for Light {
    fn reflect(&self, ray: &Ray, t: f32, normal: Vector3) -> Option<Reflection> {
        None
    }

    fn emit(&self) -> Color {
        self.color
    }
}

pub struct Metal {
    color: Color,

    _secret: (),
}

impl Metal {
    pub fn new(color: Color) -> Self {
        Self { color, _secret: () }
    }
}

impl Material for Metal {
    fn reflect(&self, ray: &Ray, t: f32, normal: Vector3) -> Option<Reflection> {
        let reflected = Vector3::reflect(ray.direction, normal);
        let reflected = Ray::new(ray.position_at(t), reflected);

        if reflected.direction.dot(normal) <= 0. {
            return None;
        }

        Some(Reflection::new(reflected, self.color))
    }

    fn emit(&self) -> Color {
        Color::black()
    }
}

pub struct Diffuse {
    color: Color,

    _secret: (),
}

impl Diffuse {
    pub fn new(color: Color) -> Self {
        Self { color, _secret: () }
    }
}

impl Material for Diffuse {
    fn reflect(&self, ray: &Ray, t: f32, normal: Vector3) -> Option<Reflection> {
        let ray = Ray::new(ray.position_at(t), Vector3::random_in_hemisphere(normal));

        Some(Reflection::new(ray, self.color))
    }

    fn emit(&self) -> Color {
        Color::black()
    }
}
